### DISCUSSION TITLE
6 useful extensions for IEnumerable

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#60] [General]
# 6 useful extensions for IEnumerable

I did already write about some useful extension methods for Task and ValueTask. Today I want to show you some useful extension methods for IEnumerable.

https://steven-giesel.com/blogPost/1b8eaaef-01a3-4799-9a96-f0d37197d175

**Comment by savemysoul382** [comment]
Hello, in your Swap method, the random type is not initialized anywhere, but is used

**Reply by linkdotnet** [reply]
The idea is that you pass in your collection you want to shuffle. The shuffle method itself doesn’t create anything more than just bringing the elements in a random order.

Maybe I am missing something. Can you explain a bit more in detail?

**Reply by jamescurran** [reply]
Swap uses an object named `random` which is never declared or initialized.  (Personally, I like putting a Random object as a singleton in my IoC container, and having it injected into my services)

**Reply by jamescurran** [reply]
Sorry, not `Swap`(), --  `Shuffle()`

**Reply by linkdotnet** [reply]
Yes I missed that. Originally it was a field in a class. I updated the implementation so it uses Random.Shared.

Much appreciated!

**Comment by maikschott** [comment]
Partitioning into two or more collections can also be done using the in-built`ToLookup` LINQ method.
```csharp
var evenOdd = numbers.ToLookup(n => n % 2 == 0);
var (even, odd) = (evenOdd[true], evenOdd[false]);
```
The `Partition` method could then likewise be shortened to:
```csharp
public static (IEnumerable<T> True, IEnumerable<T> False) Partition<T>(
        this IEnumerable<T> source,
        Func<T, bool> predicate)
{
  var partitions = source.ToLookup(predicate);
  return (partitions[true], partitions[false]);
}
```

**Reply by linkdotnet** [reply]
Ohh I like that one. Did not come into my mind. 

Yes more concise and readable!

**Reply by jamescurran** [reply]
I was going to say the you should change the return type of `Partition`  to use `IList<T>`s instead of `IEnumerable<T>`s, as the implementation use `List<>`s so there's no point in limiting what we can do with the results.

But, if you use the `ToLookup()` implementation, we have to stay with `IEnumerable<>`
